README.md
# Extra Sentence Space

A plugin for WordPress that forces browsers to display two spaces (when present) between sentences.

This plugin is available in the WordPress Plugin Directory: https://wordpress.org/plugins/extra-sentence-space/


## Installation

1. Install via the built-in WordPress plugin installer. Or install the plugin code inside the plugins directory for your site (typically `/wp-content/plugins/`).
2. Activate the plugin through the 'Plugins' admin menu in WordPress
3. Begin (or continue) to use two spaces to separate your sentences when writing a post.


## Additional Documentation

See [readme.txt](https://github.com/coffee2code/extra-sentence-space/blob/master/readme.txt) for additional usage information. See [CHANGELOG.md](CHANGELOG.md) for the list of changes for each release.


## Support

Commercial support and custom development are not presently available. You can raise an [issue](https://github.com/coffee2code/extra-sentence-space/issues) on GitHub or post in the [plugin's support forum on WordPress.org](https://wordpress.org/support/plugin/extra-sentence-space/).

If the plugin has been of benefit to you, how about [submitting a review](https://wordpress.org/support/plugin/extra-sentence-space/reviews/) for it in the WordPress Plugin Directory or considering a [donation](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6ARCFJ9TX3522)?


## License

This plugin is free software; you can redistribute it and/or modify it under the terms of the [GNU General Public License](https://www.gnu.org/licenses/gpl-2.0.html) as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

CHANGELOG.md
# Changelog

## 1.3.9 _(2019-12-31)_
* Fix: Use full path to CHANGELOG.md in the Changelog section of readme.txt
* Change: Note compatibility through WP 5.3+
* Change: Update copyright date (2020)

## 1.3.8 _(2019-06-08)_
* Change: Update unit test install script and bootstrap to use latest WP unit test repo
* Change: Note compatibility through WP 5.2+
* Change: Add link to CHANGELOG.md in README.md

## 1.3.7 _(2019-02-05)_
* New: Add CHANGELOG.md and move all but most recent changelog entries into it
* New: Add inline documentation for hook
* Change: Rename readme.txt section from 'Filters' to 'Hooks'
* Change: Add inline documentation to example in readme.txt
* Change: Split paragraph in README.md's "Support" section into two
* Change: Note compatibility through WP 5.1+
* Change: Update copyright date (2019)
* Change: Update License URI to be HTTPS

## 1.3.6 _(2018-04-14)_
* New: Add README.md
* Change: Minor whitespace tweaks to unit test bootstrap
* Change: Add GitHub link to readme
* Change: Modify formatting of hook name in readme to prevent being uppercased when shown in the Plugin Directory
* Change: Note compatibility through WP 4.9+
* Change: Update copyright date (2018)

## 1.3.5 _(2017-01-30)_
* Change: Default `WP_TESTS_DIR` to `/tmp/wordpress-tests-lib` rather than erroring out if not defined via environment variable.
* Change: Enable more error output for unit tests.
* Change: Prevent direct loading of test file.
* Change: Note compatibility through WP 4.7+.
* Change: Minor inline code documentation reformatting.
* Change: Minor readme.txt improvements.
* New: Add LICENSE file.
* Change: Update copyright date (2017).

## 1.3.4 _(2015-12-15)_
* Change: Note compatibility through WP 4.4+.
* Change: Explicitly declare methods in unit tests as public.
* Change: Update copyright date (2016).
* Add: Define 'Text Domain' header attribute.
* Add: Create empty index.php to prevent files from being listed if web server has enabled directory listings.

## 1.3.3 _(2015-08-16)_
* Update: Note compatibility through WP 4.3+
* Update: Minor documentation tweaks (spacing)

## 1.3.2 _(2015-02-11)_
* Note compatibility through WP 4.1+
* Update copyright date (2015)

## 1.3.1 _(2014-08-25)_
* Fix minor error in tests
* Minor plugin header reformatting
* Minor code reformatting (spacing)
* Change documentation links to wp.org to be https
* Note compatibility through WP 4.0+
* Add plugin icon

## 1.3 _(2013-12-14)_
* Fix bug if using '/' as custom-defined punctuation
* Add unit tests
* Note compatibility through WP 3.8+
* Update copyright date (2014)
* Add banner
* Minor readme.txt formatting tweaks
* Change donate link

## 1.2.6
* Add check to prevent execution of code if file is directly accessed
* Note compatibility through WP 3.5+
* Update copyright date (2013)

## 1.2.5
* Re-license as GPLv2 or later (from X11)
* Add 'License' and 'License URI' header tags to readme.txt and plugin file
* Remove ending PHP close tag
* Note compatibility through WP 3.4+
* Minor code reformatting (indentation)
* Update copyright date (2012)

## 1.2.4
* Note compatibility through WP 3.3+

## 1.2.3
* Note compatibility through WP 3.2+
* Tiny code formatting change (spacing)
* Fix plugin homepage and author links in description in readme.txt

## 1.2.2
* Add link to plugin homepage to description in readme.txt

## 1.2.1
* Note compatibility with WP 3.1+
* Update copyright date (2011)

## 1.2
* Add filter `c2c_extra_sentence_space_punctuation` to allow customization of the punctuation after which double-spacing (when present) is preserved. Default is '.!?'
* Add filter `c2c_extra_sentence_space` to respond to the function of the same name so that users can use the `apply_filters('c2c_extra_sentence_space')` notation for invoking function
* Wrap function in `if (function_exists())` check
* Note compatibility with WP 3.0+
* Minor code reformatting (spacing)
* Remove docs from top of plugin file (all that and more are in readme.txt)
* Remove trailing whitespace in header docs
* Add Filters and Upgrade Notice sections to readme.txt

## 1.1
* Also filter 'widget_text'
* Now also filter widget_text
* Add PHPDoc documentation
* Note compatibility with WP 2.8+ and 2.9+
* Update readme.txt (including adding Changelog)

## 1.0.1
* Note compatibility with WP 2.6+ and 2.7+
* Update copyright date
* Tweak description and extended description
* Remove commented out line of code that could be used to insert a second space if two aren't present
* Update and fix some mis-worded sentences in readme.txt

## 1.0
* Initial release


tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Extra_Sentence_Space
 */

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/tests/phpunit/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/extra-sentence-space.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/tests/phpunit/includes/bootstrap.php';


tests/test-extra-sentence-space.php
<?php

defined( 'ABSPATH' ) or die();

class Extra_Sentence_Space_Test extends WP_UnitTestCase {

	public function tearDown() {
		parent::tearDown();
		// Ensure the filter gets removed
		remove_filter( 'c2c_extra_sentence_space_punctuation', array( $this, 'more_extra_space_punctuation' ) );
	}


	//
	//
	// DATA PROVIDERS
	//
	//


	public static function two_spaces_after_handled_punctuation() {
		return array(
			array( 'This is a sentence.  Another sentence.' ),
			array( 'This is a sentence.  Another sentence.  And yet another' ),
			array( 'This is a sentence?  Another sentence.' ),
			array( 'This is a sentence!  Another sentence.' ),
		);
	}

	public static function two_spaces_after_unhandled_punctuation() {
		return array(
			array( '(This is a sentence)  Another sentence.' ),
			array( '/This is a sentence/  Another sentence.' ),
			array( '#This is a sentence#  Another sentence.' ),
			array( 'This is a sentence#  (Another sentence.)  And yet another.' ),
		);
	}

	public static function single_space_after_handled_punctuation() {
		return array(
			array( 'This is a sentence. Another sentence.' ),
			array( 'This is a sentence? Another sentence.' ),
			array( 'This is a sentence! Another sentence.' ),
		);
	}

	public static function single_space_after_unhandled_punctuation() {
		return array(
			array( '(This is a sentence) Another sentence.' ),
			array( '/This is a sentence/ Another sentence.' ),
			array( '#This is a sentence# Another sentence.' ),
		);
	}

	public static function two_spaces_after_custom_handled_punctuation() {
		return array(
			array( '(This is a sentence)  Another sentence.' ),
			array( '/This is a sentence/  Another sentence.' ),
			array( 'This is a sentence@  Another sentence.' ),
		);
	}

	public static function single_space_after_custom_handled_punctuation() {
		return array(
			array( '(This is a sentence) Another sentence.' ),
			array( '/This is a sentence/ Another sentence.' ),
			array( 'This is a sentence@ Another sentence.' ),
		);
	}

	public static function default_filters() {
		return array(
			array( 'comment_text' ),
			array( 'the_title' ),
			array( 'the_content' ),
			array( 'the_excerpt' ),
			array( 'widget_text' ),
		);
	}


	//
	//
	// FUNCTIONS FOR HOOKING ACTIONS/FILTERS
	//
	//


	public function more_extra_space_punctuation( $punctuation ) {
		// Add the '/' and ')' characters to the list of characters
		return $punctuation . '/)@';
	}

	public function preserve_extra_sentence_space( $text ) {
		return str_replace( '  ', '&nbsp; ', $text );
	}


	//
	//
	// TESTS
	//
	//


	/**
	 * @dataProvider two_spaces_after_handled_punctuation
	 */
	public function test_preserves_double_spaces_after_handled_punctuation( $text ) {
		$this->assertEquals( $this->preserve_extra_sentence_space( $text ), c2c_extra_sentence_space( $text ) );
	}

	/**
	 * @dataProvider single_space_after_handled_punctuation
	 */
	public function test_ignores_single_space_after_handled_punctuation( $text ) {
		$this->assertEquals( $text, c2c_extra_sentence_space( $text ) );
	}

	/**
	 * @dataProvider two_spaces_after_unhandled_punctuation
	 */
	public function test_ignores_double_space_after_unhandled_punctuation( $text ) {
		$this->assertEquals( $text, c2c_extra_sentence_space( $text ) );
	}

	/**
	 * @dataProvider single_space_after_unhandled_punctuation
	 */
	public function test_ignores_single_space_after_unhandled_punctuation( $text ) {
		$this->assertEquals( $text, c2c_extra_sentence_space( $text ) );
	}

	/**
	 * @dataProvider two_spaces_after_custom_handled_punctuation
	 */
	public function test_preserves_double_space_after_custom_handled_punctuation( $text ) {
		add_filter( 'c2c_extra_sentence_space_punctuation', array( $this, 'more_extra_space_punctuation' ) );
		$this->assertEquals( $this->preserve_extra_sentence_space( $text ), c2c_extra_sentence_space( $text ) );
	}

	/**
	 * @dataProvider single_space_after_custom_handled_punctuation
	 */
	public function test_ignores_single_space_after_custom_handled_punctuation( $text ) {
		add_filter( 'c2c_extra_sentence_space_punctuation', array( $this, 'more_extra_space_punctuation' ) );
		$this->assertEquals( $this->preserve_extra_sentence_space( $text ), c2c_extra_sentence_space( $text ) );
	}

	/**
	 * @dataProvider two_spaces_after_handled_punctuation
	 */
	public function test_filter_invocation( $text ) {
		$this->assertEquals( $this->preserve_extra_sentence_space( $text ), apply_filters( 'c2c_extra_sentence_space', $text ) );
	}

	/**
	 * Test that the c2c_extra_sentence_space() is applied to the filter
	 * 'the_title' as an example of it applying to handled core filters.
	 *
	 * @dataProvider two_spaces_after_handled_punctuation
	 */
	public function test_applies_to_post_title( $text ) {
		$post_id = $this->factory->post->create( array( 'post_title' => $text ) );
		$this->go_to( get_permalink( $post_id ) );
		ob_start();
		the_title();
		$title = ob_get_contents();
		ob_end_clean();
		$this->assertEquals( $this->preserve_extra_sentence_space( $text ), $title );
	}

	/**
	 * @dataProvider default_filters
	 */
	public function test_applies_to_all_default_filters( $filter ) {
		$default_hook_priority = 9;
		$this->assertEquals( $default_hook_priority, has_filter( $filter, 'c2c_extra_sentence_space' ) );
	}

}


extra-sentence-space.php
<?php
/**
 * Plugin Name: Extra Sentence Space
 * Version:     1.3.9
 * Plugin URI:  http://coffee2code.com/wp-plugins/extra-sentence-space/
 * Author:      Scott Reilly
 * Author URI:  http://coffee2code.com/
 * Text Domain: extra-sentence-space
 * License:     GPLv2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 * Description: Force browsers to display two spaces (when present) between sentences.
 *
 * Compatible with WordPress 1.5 through 5.3+.
 *
 * =>> Read the accompanying readme.txt file for instructions and documentation.
 * =>> Also, visit the plugin's homepage for additional information and updates.
 * =>> Or visit: https://wordpress.org/plugins/extra-sentence-space/
 *
* @package Extra_Sentence_Space
* @author  Scott Reilly
* @version 1.3.9
*/

/*
 * TODO:
 * - Consider adding force extra sentence space mode/setting. Under Settings -> Reading,
 *   add setting that forces a second space after punctuation. Might be
 *   tricky tho since we don't want to affect punctuation within code, for example.
 *   See: https://wordpress.org/support/topic/wasnt-working-on-all-posts/
 * - Add filter `c2c_extra_sentence_space_filters` that allows customization of
 *   the default filters affected by the plugin.
 */

/*
	Copyright (c) 2008-2020 by Scott Reilly (aka coffee2code)

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

defined( 'ABSPATH' ) or die();

if ( ! function_exists( 'c2c_extra_sentence_space' ) ) :
	/**
	 * Preserves two spaces (when present) between sentences for display as HTML.
	 *
	 * The spaces are preserved by the first space being converted to a
	 * non-breaking space entity, '&nbsp;'.
	 *
	 * @param string  $text The text to have sentence-ending double-spaces
	 *                      preserved for display as HTML.
	 * @return string       The converted text.
	 */
	function c2c_extra_sentence_space( $text ) {
		/**
		 * Filters the characters after which two spaces should be preserved. By
		 * default these are just sentence-ending punctuation.
		 *
		 * @since 1.2
		 *
		 * @param string $punctuation String of characters after which two spaces are
		 *                            preserved. Default '.!?'
		 */
		$punctuation = preg_quote( apply_filters( 'c2c_extra_sentence_space_punctuation', '.!?' ), '/' );

		return preg_replace( "/([$punctuation][\'\"]?)([ ]{2,})/imsU", "$1&nbsp; ", $text );
	}
	add_filter( 'c2c_extra_sentence_space', 'c2c_extra_sentence_space' );
endif;

add_filter( 'comment_text', 'c2c_extra_sentence_space', 9 );
add_filter( 'the_title',    'c2c_extra_sentence_space', 9 );
add_filter( 'the_content',  'c2c_extra_sentence_space', 9 );
add_filter( 'the_excerpt',  'c2c_extra_sentence_space', 9 );
add_filter( 'widget_text',  'c2c_extra_sentence_space', 9 );
